crates/ruff_python_parser/src/parser/mod.rs

```diff
@@ -62,84 +58,74 @@ impl Program {
-    /// Parse a vector of [`LexResult`]s using the specified [`Mode`].
-    pub fn parse_tokens(source: &str, tokens: Vec<LexResult>, mode: Mode) -> Program {
-        Parser::new(source, mode, TokenSource::new(tokens)).parse_program()
+    pub fn parse_starts_at(source: &str, mode: Mode, offset: TextSize) -> Program {
+        Parser::new(source, mode)
+            .with_start_offset(offset)
+            .parse_program()
     }
 }
 
 #[derive(Debug)]
 pub(crate) struct Parser<'src> {
     source: &'src str,
-    tokens: TokenSource,
+
+    /// Token source for the parser that skips over any non-trivia token.
+    tokens: TokenSource<'src>,
 
     /// Stores all the syntax errors found during the parsing.
     errors: Vec<ParseError>,
 
     /// Specify the mode in which the code will be parsed.
     mode: Mode,
 
-    /// Current token along with its range.
-    current: Spanned,
-
     /// The ID of the current token. This is used to track the progress of the parser
     /// to avoid infinite loops when the parser is stuck.
     current_token_id: TokenId,
 
-    /// The end of the last processed. Used to determine a node's end.
-    last_token_end: TextSize,
-
-    /// The range of the tokens to parse.
-    ///
-    /// The range is equal to `[0; source.len())` when parsing an entire file. The range can be
-    /// different when parsing only a part of a file using the [`crate::lex_starts_at`] and
-    /// [`crate::parse_expression_starts_at`] APIs in which case the the range is equal to
-    /// `[offset; subrange.len())`.
-    tokens_range: TextRange,
+    /// The end of the previous token processed. Used to determine a node's end.
+    prev_token_end: TextSize,
 
     recovery_context: RecoveryContext,
+
+    start_offset: TextSize,
 }
 
 impl<'src> Parser<'src> {
-    pub(crate) fn new(source: &'src str, mode: Mode, mut tokens: TokenSource) -> Parser<'src> {
-        let tokens_range = TextRange::new(
-            tokens.position().unwrap_or_default(),
-            tokens.end().unwrap_or_default(),
-        );
+    pub(crate) fn new(source: &'src str, mode: Mode) -> Parser<'src> {
+        let mut tokens = TokenSource::from_source(source, mode);
 
-        let current = tokens
-            .next()
-            .unwrap_or_else(|| (Tok::EndOfFile, TextRange::empty(tokens_range.end())));
+        // Initialize the token source so that the current token is set correctly.
+        tokens.next_token();
 
         Parser {
             mode,
             source,
             errors: Vec::new(),
             tokens,
             recovery_context: RecoveryContext::empty(),
-            last_token_end: tokens_range.start(),
-            current,
+            prev_token_end: TextSize::new(0),
+            start_offset: TextSize::new(0),
             current_token_id: TokenId::default(),
-            tokens_range,
         }
     }
 
+    pub(crate) fn with_start_offset(mut self, offset: TextSize) -> Parser<'src> {
+        self.start_offset = offset;
```

Comment:
> Related. Do we need to offset the tokens by `start_offset`? I think the current `lex_starts_at` doesn't do that but I think it's odd if the AST nodes are all offseted by `start_offset` and the tokens arent. Or is this also not the case?

Yes, that is true. I missed that.

> I think the way I would expect this to work is that we pass `start_offset` right through to the lexer and the lexer sets its current position to `start_offset`.

Yes, it should be passed directly to the `Lexer` which will set the correct range for the tokens and then the parser can just use the range.

I'll still store the offset on the parser to correctly extract the source code for `src_text` function.

Reply:
> Do we need to update `prev_token_end` as well? I think I would make this a constructor function as well to avoid any other invariants

Yeah, that makes sense. That should be updated as well, I'll make it a separate constructor function.

Reply:
Fixed in #11433 

I'm going to test this out in the local fork of the parser to make sure the ranges are correct.